src/sdc11073/wsdiscovery/wsdimpl.py

```diff
@@ -0,0 +1,620 @@
+from __future__ import annotations
+
+import logging
+import platform
+import random
+import time
+from typing import TYPE_CHECKING, Callable
+from urllib.parse import unquote, urlsplit
+
+from sdc11073.definitions_sdc import SDC_v1_Definitions
+from sdc11073.exceptions import ApiUsageError
+from sdc11073.namespaces import default_ns_helper as nsh
+from sdc11073.netconn import get_ip_for_adapter, get_ipv4_addresses, get_ipv4_ips
+from sdc11073.xml_types import wsd_types
+from sdc11073.xml_types.addressing_types import HeaderInformationBlock
+
+from .common import MULTICAST_IPV4_ADDRESS, MULTICAST_PORT, message_factory
+from .networkingthread import NetworkingThreadPosix, NetworkingThreadWindows
+from .service import Service
+
+if TYPE_CHECKING:
+    from logging import Logger
+
+    from lxml.etree import QName
+
+    from sdc11073.pysoap.msgreader import ReceivedMessage
+    from sdc11073.pysoap.msgfactory import CreatedMessage
+    from sdc11073.xml_types.msg_types import MessageType
+    from sdc11073.location import SdcLocation
+    from collections.abc import Iterable
+
+
+WSA_ANONYMOUS = nsh.WSA.namespace + '/anonymous'
+ADDRESS_ALL = "urn:docs-oasis-open-org:ws-dd:ns:discovery:2009:01"  # format acc to RFC 2141
+
+NS_D = nsh.WSD.namespace
+MATCH_BY_LDAP = NS_D + '/ldap'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/ldap"
+MATCH_BY_URI = NS_D + '/rfc3986'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/rfc3986"
+MATCH_BY_UUID = NS_D + '/uuid'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/uuid"
+MATCH_BY_STRCMP = NS_D + '/strcmp0'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/strcmp0"
+
+
+def types_info(types: list[QName]) -> list[str]:
+    """Make printable strings from list of QNames (helper method for logging)."""
+    return [str(t) for t in types] if types else types
+
+
+def match_scope(my_scope: str, other_scope: str, match_by: str) -> bool:
+    """match_scope checks if my_scope matches other_scope by applying the algorithm defined by match_by.
+
+    match_scope correctly handles "%2F" (== '/') encoded values.
+    """
+    if match_by in (MATCH_BY_LDAP, MATCH_BY_URI, MATCH_BY_UUID, '', None):

```

Comment:
what is `''` and `None` for a matching algorithm? `None` is not really an applicable value, because `match_by` is never `None` because of its type

Reply:
I dont know either, I saw this in an example years ago.